### PR TITLE
Adding support for BlockedKV attention in CasualLM models

### PR DESCRIPTION
**Pull request by vaibverm**
### Objective:

This PR introduces the KV blocking technique for CausalLM models where the K/V cache is read and processed block by block in the attention computation. Number of desired KV blocks are defined at model initialization in the "from_pretrained" call to export the ONNX with required number of KV blocks. As a result, the following changes are introduced:

### Changes:
1. SoftMax needs to be changed from regular SoftMax to online SoftMax where the running maximum and cumulative denominators are tracked and updated once each block is processed to retain mathematical accuracy compared to regular SoftMax.
2. Changes to CTXGather and CTXGatherCB custom ops to read only 1 block worth of data in each cache gather/read.
3. Changes to read_only function in QEffDynamicCache to allow reading of a cache block by block rather than full K/V cache.
4. Generation of attention mask per block.
5. Changes to eager_attention_forward implementation in the llama model to allow BlockedKV attention and online SoftMax implementation. 
6. Wrapping the num_kv_blocks variable inside qaic_config to keep consistent calling style. 
7. A new PyTorch transform to pass the num_kv_blocks variable to QEffLlamaAttention block. 
8. A new constant added for num_kv_blocks.
9. Added tests to switch the BlockedKV feature on and off.

Please review and feel free to suggest changes and tests.

### CODE REVIEW

**Comment by vaibverm** [comment]
Hi @vbaddi, 
I have addressed the conflicts but some workflows need approval. Would you be able to approve those?

File: QEfficient/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -103,21 +103,95 @@ def eager_attention_forward(
...
-    attn_weights = torch.matmul(query, key_states.transpose(2, 3)) * scaling
-    if attention_mask is not None:
-        attn_weights = torch.where(
-            attention_mask, torch.tensor(MIN_MASKED_ATTENTION_VALUE, dtype=torch.float32), attn_weights
+    if num_kv_blocks is not None:
+        # Initialize result tensor
+        output = torch.zeros_like(query)
+
+        # Perform blockwise computation for K and V
+        M = torch.full(
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/llama/modeling_llama.py]: better to define batch_size, num_heads, seq_len, _ = query.shape and reuse it, for better redablity, also use proper variable naming

File: QEfficient/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -103,21 +103,95 @@ def eager_attention_forward(
...
+        position_ids = cache_kwargs.get("position_ids")
+        block_size = -(-past_seen_tokens // num_kv_blocks)
+        for j in range(num_kv_blocks):
+            start_index = j * block_size
+            end_index = (j + 1) * block_size
+            K_block, V_block = past_key_value.read_only_blockedKV(start_index, end_index, layer_idx, cache_kwargs)
+            K_block_states = repeat_kv(K_block, module.num_key_value_groups)
+            V_block_states = repeat_kv(V_block, module.num_key_value_groups)
+            past_seen_tokens_start = start_index
+            past_seen_tokens_end = torch.where(
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/llama/modeling_llama.py]: if we are comparing int do we need torch.where? can we use min()?

File: QEfficient/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -103,21 +103,95 @@ def eager_attention_forward(
...
+                torch.tensor(past_seen_tokens, dtype=torch.int) < torch.tensor(end_index, dtype=torch.int),
+                past_seen_tokens,
+                end_index,
+            )
+            causal_mask_block = _create_causal_mask(
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/llama/modeling_llama.py]: move the causal_mask_block inside the if attention_mask is not None: condition

File: QEfficient/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -103,21 +103,95 @@ def eager_attention_forward(
...
+
+            # Compute attention scores for the block
+            attn_weights_block = torch.matmul(query, K_block_states.transpose(2, 3)) * scaling
+            if attention_mask is not None:
+                attn_weights_block = torch.where(
+                    causal_mask_block, torch.tensor(MIN_MASKED_ATTENTION_VALUE, dtype=torch.float32), attn_weights_block
+                )
+
+            # Update Running row maximum
+            prevM = M
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/llama/modeling_llama.py]: Do use snake case for all variable and use proper variable names

File: QEfficient/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -103,21 +103,95 @@ def eager_attention_forward(
...
+            # update Running denominator
+            prevD = D
+            D = prevD * torch.exp(deltaM) + currentExp.sum(axis=-1)
+
+            P = currentExp / D.unsqueeze(-1)
+
+            prevO = output
+            output = ((prevD / D).unsqueeze(-1)) * prevO * torch.exp(deltaM.unsqueeze(-1)) + torch.matmul(
+                P, V_block_states
+            )  # This in higher precision.
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/llama/modeling_llama.py]: is the comment "This in higher precision" needed?

File: QEfficient/transformers/models/llama/modeling_llama.py [hunk 6]
```diff
@@ -151,31 +226,44 @@ def forward(
...
+                }
+                past_key_value.write_only(key_states, value_states, self.layer_idx, cache_kwargs)
+            else:
+                cache_kwargs = {"batch_index": batch_index, "position_ids": position_ids}
+                if comp_ctx_lengths is not None:
+                    attention_mask = attention_mask[:, :, :, : comp_ctx_lengths.shape[-1]]
+                    cache_kwargs["CCL"] = attention_mask.shape[-1]
+                key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
 
         attention_interface = eager_attention_forward
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/llama/modeling_llama.py]: Can we have a condition here to switch between the attention forward if num_kv_blocks is passed? we can pass the num_kv_blocks=num_kv_blocks, position_ids layer idx and pkv as kwargs

File: QEfficient/transformers/models/modeling_auto.py [hunk 7]
```diff
@@ -2306,6 +2308,14 @@ def __init__(
         if self.is_tlm:
             self.model.qaic_config["return_pdfs"] = True
 
+        if self.model.qaic_config is not None:
+            num_kv_blocks = self.model.qaic_config.get("num_kv_blocks", None)
+        else:
+            num_kv_blocks = None
+
+        if num_kv_blocks is not None:
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/modeling_auto.py]: can we move it line 2313? 
<if self.model.qaic_config is not None and self.model.qaic_config.get("num_kv_blocks", None) is not None:
    BlockedKVAttentionTransform.apply(model, num_kv_blocks=self.model.qaic_config.get("num_kv_blocks")>

File: QEfficient/transformers/cache_utils.py [hunk 8]
```diff
@@ -263,6 +308,25 @@ def read_only(self, layer_idx, cache_kwargs):
         """
         return self.layers[layer_idx].read_only(cache_kwargs)
 
+    def read_only_blockedKV(self, start_index, end_index, layer_idx, cache_kwargs):
```
Comment (quic-rishinr) [inline QEfficient/transformers/cache_utils.py]: same as above

File: QEfficient/transformers/cache_utils.py [hunk 9]
```diff
@@ -61,6 +63,49 @@ def read_only(self, cache_kwargs):
             v_out = CtxGatherFunc.apply(v_out, ctx_indices, ctx_len)
 
         v_out = torch.where(invalid_mask.unsqueeze(-1), torch.tensor(0.0, dtype=torch.float32), v_out)
+
+    def read_only_blockedKV(self, start_index, end_index, cache_kwargs):
```
Comment (quic-rishinr) [inline QEfficient/transformers/cache_utils.py]: can we repurpose the read_only method by providing default values to start_index and end_index? start idx can be 0 and end idx can be ctx length

File: QEfficient/transformers/models/llama/modeling_llama.py [hunk 10]
```diff
@@ -103,21 +103,95 @@ def eager_attention_forward(
...
     **kwargs,
 ):
-    key_states = repeat_kv(key, module.num_key_value_groups)
-    value_states = repeat_kv(value, module.num_key_value_groups)
-
-    attn_weights = torch.matmul(query, key_states.transpose(2, 3)) * scaling
-    if attention_mask is not None:
-        attn_weights = torch.where(
-            attention_mask, torch.tensor(MIN_MASKED_ATTENTION_VALUE, dtype=torch.float32), attn_weights
+    if num_kv_blocks is not None:
```
[the remark this answers is not in the document]
Comment (vaibverm) [reply QEfficient/transformers/models/llama/modeling_llama.py]: Created separate methods for blockedKV attention and regular eager attention.

File: QEfficient/transformers/models/llama/modeling_llama.py (hunk 4 above)
Comment (vaibverm) [reply QEfficient/transformers/models/llama/modeling_llama.py]: Done

File: QEfficient/transformers/models/llama/modeling_llama.py [hunk 11]
```diff
@@ -103,21 +103,95 @@ def eager_attention_forward(
...
+            )
+            causal_mask_block = _create_causal_mask(
+                position_ids=position_ids, target_length=past_seen_tokens_end, start_index=past_seen_tokens_start
+            )
+
+            # Compute attention scores for the block
+            attn_weights_block = torch.matmul(query, K_block_states.transpose(2, 3)) * scaling
+            if attention_mask is not None:
+                attn_weights_block = torch.where(
+                    causal_mask_block, torch.tensor(MIN_MASKED_ATTENTION_VALUE, dtype=torch.float32), attn_weights_block
```
[the remark this answers is not in the document]
Comment (vaibverm) [reply QEfficient/transformers/models/llama/modeling_llama.py]: I have done this now but didn't do it earlier since this was used just once within the loop.

File: QEfficient/transformers/models/llama/modeling_llama.py [hunk 12]
```diff
@@ -103,21 +103,95 @@ def eager_attention_forward(
...
+                )
+
+            # Update Running row maximum
+            prevM = M
+            M = torch.max(prevM, torch.max(attn_weights_block, axis=-1).values)
```
[the remark this answers is not in the document]
Comment (vaibverm) [reply QEfficient/transformers/models/llama/modeling_llama.py]: I have updated the variable naming and using attn_weights_block.max(dim=-1).values now. Please let me know if readability is still an issue.

File: QEfficient/transformers/models/llama/modeling_llama.py (hunk 5 above)
Comment (vaibverm) [reply QEfficient/transformers/models/llama/modeling_llama.py]: Removed the comment. Was in there based on legacy discussions.

File: QEfficient/transformers/models/llama/modeling_llama.py (hunk 6 above)
Comment (vaibverm) [reply QEfficient/transformers/models/llama/modeling_llama.py]: I have updated with the condition. Please check to see if it resolves the concerns. Kindly let me know if any further changes are needed.

File: QEfficient/transformers/cache_utils.py (hunk 8 above)
Comment (vaibverm) [reply QEfficient/transformers/cache_utils.py]: The non-blocked read_only method is being used in modeling_llama_swiftkv.py as well. That's the reason I created a new version.

File: QEfficient/transformers/models/pytorch_transforms.py [hunk 13]
```diff
@@ -821,3 +822,20 @@ def apply(cls, model: nn.Module, pooling: Union[str, Callable]) -> Tuple[nn.Modu
...
+
+    @classmethod
+    def apply(cls, model: nn.Module, num_kv_blocks) -> Tuple[nn.Module, bool]:
+        transformed = False
+        for module in model.modules():
+            if type(module) in cls._module_mapping:
+                repl_module = type(module)
+                module.__class__ = repl_module
+                module.forward = MethodType(partial(repl_module.forward, num_kv_blocks=num_kv_blocks), module)
+                transformed = True  # Set to True if at least one transformation occurs
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/transformers/models/pytorch_transforms.py]: it would be better to do it with pytorch transform

[the remark this answers is not in the document]
Comment (vaibverm) [reply QEfficient/transformers/models/pytorch_transforms.py]: Added the warning for unsupported architectures.
